### PR TITLE
feat(notifier): add ability to contribute to the website by uploading patch files

### PR DESCRIPTION
**Pull request by roj1512**
This PQ adds a functionality to the Notifier bot that lets users in the configured chat push commits to specific branches of the repository.

1. The user sends a patch file to the configured chat, with a caption that refers to the branch the user wants to push their patch to.
2. The bot sends uploads the patch to an instance of [patch-pusher](https://github.com/roj1512/patch-pusher).
3. The patch is put on a queue to be committed and pushed to the specified branch.

Other notable changes:

- The `Bot` now listens to updates on `/<BOT_TOKEN>`.

Benefits of this:

- This lets users contribute to the website translation without the need of having a GitHub account.

### CODE REVIEW

File: notifier/main.ts [hunk 1]
```diff
@@ -1,12 +1,41 @@
...
+
+bot.chatType(["group", "supergroup"]).filter((ctx) =>
+  ctx.chat.id == env.CHAT_ID
+).on("message:document", async (ctx) => {
+  if (
+    !ctx.message.document.file_name?.endsWith(".patch") || !ctx.message.caption
+  ) {
+    return;
+  }
+  const file = await fetch((await ctx.getFile()).getUrl());
```
Comment (KnorpelSenf) [inline notifier/main.ts:23]: Doesn't this load the entire file in memory?

Comment (roj1512) [reply notifier/main.ts:23]: No, it doesn’t, It’s only loading the response status and headers in memory. The actual content is later piped to another request as blob:

https://github.com/grammyjs/website/blob/feat-notifier---add-ability-to-contribute-to-the-website-by-uploading-patch-files/notifier/main.ts#L27

Comment (KnorpelSenf) [reply notifier/main.ts:23]: Seems like it does: https://developer.mozilla.org/en-US/docs/Web/API/Response/blob

Comment (roj1512) [reply notifier/main.ts:23]: Hmm.

**Comment by KnorpelSenf** [comment]
Will review tomorrow. What happened to your branch name?

**Comment by roj1512** [comment]
What happened to my branch name?

**Comment by KnorpelSenf** [comment]
It's so long that when I started reading it, I was interrupted midway because my sister was faster watching all Lord of the Rings movies than I was reading the first half.

**Comment by roj1512** [comment]
You’re welcome.

**Comment by roj1512** [comment]
I used https://github.com/roj1512/b.

**Comment by KnorpelSenf** [comment]
Wait so you copy in a branch name and it intercalates hyphens?

**Comment by roj1512** [comment]
A commit message, not PQ name.

**Comment by KnorpelSenf** [comment]
> It's so long that when I started reading it, I was interrupted midway because my sister was faster watching all Lord of the Rings movies than I was reading the first half.

The perfect response would've been, “I also once knew someone who read that slowly” 😜

File: notifier/import_map.json [hunk 2]
```diff
@@ -1,10 +1,11 @@
 {
   "imports": {
     "envalid/": "https://deno.land/x/envalid@0.1.2/",
-    "grammy/": "https://deno.land/x/grammy@v1.11.2/",
+    "grammy/": "https://deno.land/x/grammy@v1.13.1/",
```
Comment (KnorpelSenf) [inline notifier/import_map.json:4]: Consider updating now that we have a new version

File: notifier/import_map.json [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
     "oak/": "https://deno.land/x/oak@v11.1.0/",
-    "octokit/": "https://raw.githubusercontent.com/octokit/webhooks-methods.js/v2.0.0/src/",
+    "octokit-webhooks-methods": "https://esm.sh/@octokit/webhooks-methods@3.0.2",
     "postgres/": "https://deno.land/x/postgres@v0.17.0/",
-    "std/": "https://deno.land/std@0.164.0/"
+    "std/": "https://deno.land/std@0.175.0/",
```
Comment (KnorpelSenf) [inline notifier/import_map.json:8]:
```suggestion
    "std/": "https://deno.land/std@0.176.0/",
```

File: notifier/import_map.json (hunk 2 above)
Comment (KnorpelSenf) [reply notifier/import_map.json:4]:
```suggestion
    "grammy/": "https://deno.land/x/grammy@v1.14.0/",
```

File: notifier/main.ts [hunk 4]
```diff
@@ -1,12 +1,67 @@
...
+  }
+  const repliedMessage = ctx.message.reply_to_message;
+  if (
+    !repliedMessage || !(repliedMessage.from?.id == ctx.me.id)
+  ) {
+    return;
+  }
+  const prNumber = Number(
+    (repliedMessage.text ?? "").match(/^#([1-9][0-9]+)\s/)?.[1],
+  );
```
Comment (KnorpelSenf) [inline notifier/main.ts:31]: How about parsing this from the message entities in order to avoid the regex?

File: notifier/main.ts [hunk 5]
```diff
@@ -1,12 +1,67 @@
 import * as path from "std/path/mod.ts";
-import { Application } from "oak/mod.ts";
-import { Bot } from "grammy/mod.ts";
+import { Application, Router } from "oak/mod.ts";
+import { Bot, Context, webhookCallback } from "grammy/mod.ts";
+import { FileFlavor, hydrateFiles } from "grammy_files/mod.ts";
```
Comment (KnorpelSenf) [inline notifier/main.ts:4]:
```suggestion
import { Bot, type Context, webhookCallback } from "grammy/mod.ts";
import { type FileFlavor, hydrateFiles } from "grammy_files/mod.ts";
```

Comment (roj1512) [reply notifier/main.ts:4]: It doesn’t make a difference when compiled, but OK.

File: notifier/main.ts (hunk 4 above)
Comment (roj1512) [reply notifier/main.ts:31]: Alright.

File: notifier/main.ts [hunk 6]
```diff
@@ -26,8 +26,12 @@ bot.chatType(["group", "supergroup"]).filter((ctx) =>
   ) {
     return;
   }
+  const entity = repliedMessage.entities?.[0];
```
Comment (KnorpelSenf) [inline notifier/main.ts:29]: Confer https://deno.land/x/grammy@v1.14.0/mod.ts?s=Context&p=prototype.entities

Comment (roj1512) [reply notifier/main.ts:29]: Like this?

```suggestion
  const entity = ctx.entities("text")?.[0];
```

Comment (KnorpelSenf) [reply notifier/main.ts:29]: Yes, but without `?.`.

Comment (KnorpelSenf) [reply notifier/main.ts]: Oh and also, the name of the entity isn't `"text"`. It's either `"url"` or `"text_link"`.

File: notifier/main.ts [hunk 7]
```diff
@@ -1,12 +1,71 @@
...
+  ) {
+    return;
+  }
+  const repliedMessage = ctx.message.reply_to_message;
+  if (
+    !repliedMessage || !(repliedMessage.from?.id == ctx.me.id)
+  ) {
+    return;
+  }
+  const entity = ctx.entities("text")[0];
```
Comment (roj1512) [inline notifier/main.ts:29]:
```suggestion
  const entity = ctx.entities("text_link")[0];
```

**Review by KnorpelSenf — APPROVED** [review]
Rubber-stamp LGTM.

**Pull request merged**